Add documentation about deprecating Leap features (#377)
dwavesystems/dwave-ocean-sdk: docs/leap_sapi/sapi_rest.rst

    :ref:`error responses <sapi_rest_general_response_codes>`, SAPI may return
    particular error codes.

.. _leap_sapi_deprecation_message:

SAPI Deprecation Messages
-------------------------

During the deprecation period, a deprecation message is sent when such a
feature is used. For example, a message is sent when you use a deprecated
solver parameter or submit a problem to a deprecated solver.

Deprecation messages are sent in SAPI responses via the HTTP ``X-Deprecation``
response header, which adheres to

> `RFC 9651: Structured Field Values for HTTP <https://www.rfc-editor.org/rfc/rfc9651>`_.

Reviewer: [RFC9651](https://datatracker.ietf.org/doc/rfc9651/) states that to "specify an HTTP field as a Structured Field, its authors need to..." and gives points you need to meet and shows an example of how to do that. Do we want to write this more in accordance to that, especially w.r.t. specifying the type of field values?

```
The HTTP X-Deprecation response header conveys information about 
deprecated SAPI  features. It is an Item Structured Header Field 
[RFC9651](https://www.rfc-editor.org/rfc/rfc9651). Its value is a String 
(Section 3.3.3 of [RFC9651]).  Its value is a deprecation ID that may be used
to find a relevant [release note](https://docs.dwavequantum.com/projects/leap_sapi/en/latest/release_notes.html).

The following parameters are defined:

*  A parameter whose key is "context", and whose value is a String (Section 3.3.3 of [RFC9651]), 
   conveys information about the affected Leap service category as follows:

   -  "api": SAPI, such as endpoints, HTTP headers, and data structures
```

and so on, similar to example 2.1 of a header specification. The reason being that the RFC is to help parsers by strictly defining the types.

Author: Does a developer need to know this much?

A deprecation message includes the following:

*   ``context`` (string): which identifies the affected Leap service category as
    follows:

    *   ``api``: SAPI, such as endpoints, HTTP headers, and data structures

    *   ``feature``: Solver features

    *   ``parameter``: Solver parameters

    *   ``solver``: QPU and hybrid solvers

*   ``deprecated`` (date, optional): Date the feature was deprecated

> *   ``deprecation-id`` (string): Deprecation ID, which can be used to find the

Author: It's `dep-1`; I clarified that.

    relevant release note

*   ``link`` (string, optional): URI to the release note for the deprecation

*   ``message`` (string): Brief description

> *   ``sunset`` (date): Sunset date, after which support for the feature is no

Reviewer 2: May be useful to mention the date format?
It's unix timestamp

Author: Added a link to date in the RFC.

    longer guaranteed

An example of a deprecation message is the following:

.. code-block::

    x-deprecation:
        dep-1;

>             message="Missing Accept header or unsupported content type in Accept Header. Please use the allowed content type for version 3 in Accept header.";

Reviewer: Why indent the parameters more than the "dep-1" indentation?

<img width="145" height="157" alt="image" src="https://github.com/user-attachments/assets/c1870f4f-d7d6-4ce4-a029-6f5884cc5ea6" />

Is this to distinguish between Item and parameters?

            sunset=@1717171200;
            context="api";
            deprecated=@1714531200;

>             link="https://docs.dwavesystems.com/projects/leap_sapi/en/latest/release_notes.html"

Reviewer 3: cosmetically it's fine to show in multi-line; but IIRC the HTTP Spec headers + values are all in a single line. No change require, just pointing this out

Reviewer 3: suggestion: should we order the fields alphabetically to align with the order of the field described above? or based on usefulness (eg. context, message, deprecated, sunset, link)

Author: It's easier to scan if a list is sorted alphabetically.

    follows:

    *   ``api``: SAPI, such as endpoints, HTTP headers, and data structures

    *   ``feature``: Solver features

    *   ``parameter``: Solver parameters

    *   ``solver``: QPU and hybrid solvers

*   ``deprecated`` (date, optional): `Date <https://www.rfc-editor.org/rfc/rfc9651#name-dates>`_
    the feature was deprecated

*   ``deprecation-id`` (string): Deprecation ID, which directly follows
    ``x-deprecation`` and can be used to find the relevant release note

*   ``link`` (string, optional): URI to the release note for the deprecation

*   ``message`` (string): Brief description

*   ``sunset`` (date): Sunset `date <https://www.rfc-editor.org/rfc/rfc9651#name-dates>`_,
    after which support for the feature is no longer guaranteed

An example of a deprecation message\ [#]_ is the following, where ``dep-1`` is
the ``deprecation-id``:

.. code-block::

    x-deprecation:
        dep-1;
            message="Missing Accept header or unsupported content type in Accept Header. Please use the allowed content type for version 3 in Accept header.";
            sunset=@1717171200;
            context="api";
            deprecated=@1714531200;

>             link="https://docs.dwavesystems.com/projects/leap_sapi/en/latest/release_notes.html"

Author: @randomir -- `link` only goes to the top-level of the release notes so that we don't have to remember to change the link for every new deprecation.

    :ref:`error responses <sapi_rest_general_response_codes>`, SAPI may return
    particular error codes.

.. _leap_sapi_deprecation_message:

SAPI Deprecation Messages
-------------------------

The Leap service sends deprecation messages when you use deprecated features;
for example, a message is sent when you use a deprecated solver parameter or
submit a problem to a deprecated solver.

Deprecation messages are sent in SAPI responses via an HTTP ``X-Deprecation``

> response header, which is an Item Structured Header field

Reviewer: ```suggestion
response header, which is an List Structured Header field
```

(`RFC 9651 <https://www.rfc-editor.org/rfc/rfc9651>`_). Thus, a deprecation
message is a
List (`RFC 9651, 3.1 <https://www.rfc-editor.org/rfc/rfc9651#name-lists>`_)

> of Token
> (`RFC 9651, 3.3.4 <https://www.rfc-editor.org/rfc/rfc9651#name-tokens>`_)
> items, where each item is associated with the following

Reviewer: ```suggestion
of Tokens
(`RFC 9651, 3.3.4 <https://www.rfc-editor.org/rfc/rfc9651#name-tokens>`_),
each associated with the following
```

To avoid having "Item (RFC, section)"

Parameters
(`RFC 9651, 3.1.2 <https://www.rfc-editor.org/rfc/rfc9651#name-parameters>`_):

*   ``context`` (String
    (`RFC 9651, 3.3.3 <https://www.rfc-editor.org/rfc/rfc9651#name-strings>`_)):

>     which identifies the affected Leap service category as follows:

Reviewer: ```suggestion
    identifies the affected Leap service category with one of the following values:
```

    *   ``api``: SAPI, such as endpoints, HTTP headers, and data structures

    *   ``feature``: Solver features

    *   ``parameter``: Solver parameters

    *   ``solver``: QPU and hybrid solvers

*   ``deprecated`` (Date
    (`RFC 9651, 3.3.7 <https://www.rfc-editor.org/rfc/rfc9651#name-dates>`_),
    optional): Date that the feature was deprecated

> *   ``deprecation-id`` (Token

Reviewer: And then I think this would be separate from the Parameters
